### PR TITLE
Support Ethereum-compatible accounts and withdrawal from Emerald

### PR DESCRIPTION
**Pull request by lvshaoping007**
1. add Ethereum-compatible Account support
2. add emerald withdraw support
3. fix some transaction name

### CODE REVIEW

**Review by pro-wh — COMMENTED** [review]
I've had a quick look through. We should test the deposit and withdraw on our side.

File: src/background/service/APIService.js
Comment (lvshaoping007) [reply src/background/service/APIService.js:369]: What does this mean, can you explain it?
there change because when wallet have evmAddress , account will add evmAddress
[this link](https://github.com/oasisprotocol/oasis-wallet-ext/blob/b5b69d732a2c0c345a1be2db5fc30ff7beb62bb3/src/background/service/APIService.js#L377)

File: src/popup/pages/AccountInfo/index.js
Comment (lvshaoping007) [reply src/popup/pages/AccountInfo/index.js:51]: Yes, it was not deleted when the UI was updated before

File: src/popup/pages/AccountManage/index.js
Comment (lvshaoping007) [reply src/popup/pages/AccountManage/index.js]: After putting together some new ideas, I will update these

File: src/popup/pages/Main/index.js
Comment (lvshaoping007) [reply src/popup/pages/Main/index.js:39]: Ah, this is the test code, I forgot to delete it . I will update this later

Comment (lvshaoping007) [reply src/popup/pages/Main/index.js]: fixed this in near [commit](https://github.com/oasisprotocol/oasis-wallet-ext/pull/168/commits/04ed8e8f722324e3964c9f41f72a0229bdf184cb)

File: src/popup/pages/AccountManage/index.js
Comment (lvshaoping007) [reply src/popup/pages/AccountManage/index.js]: fixed this in this [commit](https://github.com/oasisprotocol/oasis-wallet-ext/pull/168/commits/b2c7c6b948ff82c2c873d45ebc10184b870d1fad)

File: src/background/service/APIService.js
Comment (pro-wh) [reply src/background/service/APIService.js:369]: https://jsfiddle.net/afeb9m13/

see this sample snippet. you should be able to set `.evmAddress` on a `const`

File: src/popup/pages/AccountManage/index.js
Comment (pro-wh) [reply src/popup/pages/AccountManage/index.js]: thanks!

File: src/popup/pages/Main/index.js
Comment (pro-wh) [reply src/popup/pages/Main/index.js]: oh ok, great

**Comment by pro-wh** [comment]
couple of notes from testing a local build:

1. clicking the button to copy the ethereum compatible address copies the oasis address instead
2. the jazzicon on an imported account doesn't match metamask's rendition of the same account. seems not critical, but I can't help but look into it

File: src/popup/pages/AccountManage/index.js
Comment (lukaw3d) [reply src/popup/pages/AccountManage/index.js:156]: Could do 
```
className="evmAccountTitle"
className="oasisAccountTitle"
.evmAccountTitle, .oasisAccountTitle {
```

File: src/popup/component/AccountIcon/index.js
Comment (lukaw3d) [reply src/popup/component/AccountIcon/index.js:24]:
```suggestion
    // Compatible with https://github.com/MetaMask/metamask-extension/blob/v10.7.0/ui/helpers/utils/icon-factory.js#L84-L88
    jsNumberForAddress(address) {
        const addr = address.slice(2, 10);
        const seed = parseInt(addr, 16);
        return seed;
    }
```

File: package.json
Comment (pro-wh) [inline package.json:45]: :information_source: this package was already in our dependency tree through web3-utils (we use its `isAddress`)

File: package.json
Comment (pro-wh) [inline package.json:17]: we should commit the corresponding yarn.lock changes too

File: src/popup/pages/AccountInfo/index.js
Comment (pro-wh) [inline src/popup/pages/AccountInfo/index.js:249]: here it uses `copyAddress`, which copies the oasis address.

File: src/popup/component/AccountIcon/index.js
Comment (pro-wh) [inline src/popup/component/AccountIcon/index.js:17]: here `address` is always the oasis address. we'd need to pass in the evmAddress from outside to use jsNumberForAddress.

<details>

```diff
diff --git a/src/popup/component/AccountIcon/index.js b/src/popup/component/AccountIcon/index.js
index 113b7d2..6503bae 100644
--- a/src/popup/component/AccountIcon/index.js
+++ b/src/popup/component/AccountIcon/index.js
@@ -9,12 +9,12 @@ export default class AccountIcon extends PureComponent {
         this.appendJazzicon()
     }
     appendJazzicon() {
-        const { address, diameter } = this.props
-        const image = this.generateJazziconSvg(address, diameter)
+        const { address, evmAddress, diameter } = this.props
+        const image = this.generateJazziconSvg(address, evmAddress, diameter)
         this.iconContainer.current.appendChild(image)
     }
-    generateJazziconSvg = (address, diameter) => {
-        const numericRepresentation = isEvmAddress(address) ? this.jsNumberForAddress(address):this.addressToNumber(address)
+    generateJazziconSvg = (address, evmAddress, diameter) => {
+        const numericRepresentation = evmAddress ? this.jsNumberForAddress(evmAddress):this.addressToNumber(address)
         return jazzicon(diameter, numericRepresentation)
     }
     jsNumberForAddress(address) {
diff --git a/src/popup/component/WalletBar/index.js b/src/popup/component/WalletBar/index.js
index d226695..f0d35f2 100644
--- a/src/popup/component/WalletBar/index.js
+++ b/src/popup/component/WalletBar/index.js
@@ -110,9 +110,11 @@ class WalletBar extends React.Component {
     renderMyIcon = () => {
         let { currentAccount } = this.props
         let address = currentAccount.address
+        let evmAddress = currentAccount.evmAddress
         return (
             <AccountIcon
                 address={address}
+                evmAddress={evmAddress}
                 diameter={"30"}
             />
         )
```

</details>

File: src/background/service/APIService.js
Comment (lukaw3d) [reply src/background/service/APIService.js:369]: (I actually like leaving variables with deep mutations as `let` too :)

File: src/background/api/index.js
Comment (lukaw3d) [inline src/background/api/index.js:219]: if propDecimals = 0, then decimals should = 0
```suggestion
  const decimals = propDecimals ?? 9
```

Comment (lukaw3d) [reply src/background/api/index.js:219]:
```suggestion
  lconst decimals = propDecimals ??  cointypes.decimals
```

File: src/popup/component/AccountIcon/index.js
Comment (pro-wh) [reply src/popup/component/AccountIcon/index.js:17]: oh there are some other places that use `<AccountIcon>` that also would need to be updated

**Comment by pro-wh** [comment]
Ok I've tried doing a withdraw (and deposit again) on testnet. Some observations:

3. I was switching around accounts to get all the addresses I need copied, and at one point the paratime tab showed "unknown" for the paratime names

And some non-critical ones:

4. We should put a space between the paratime name and the paratime ID
5. It would be convenient if you could copy the evm address from the wallet tab, without having to go into the account details screen. you can currently only copy the oasis address this way.
6. The balance of paratime accounts is available when you go to withdraw. It might be nice to have it in the paratime tab too. But the Oasis side would probably like to discuss the design of this first.

File: src/background/service/APIService.js
Comment (lvshaoping007) [reply src/background/service/APIService.js:369]: Can you talk about why you want to change to `const`? For variable parameters, I prefer to use `let`

File: src/popup/pages/AccountManage/index.js
Comment (lvshaoping007) [reply src/popup/pages/AccountManage/index.js:156]: ok , i will fix this

File: src/background/service/APIService.js
Comment (pro-wh) [reply src/background/service/APIService.js:369]: it enforces that later changes to the code doesn't accidentally reassign the variable. but your guess is as good as mine whether we'll first encounter that or we'll first want to reassign it for real. feel free to keep it this way, given that you prefer this.

File: package.json
Comment (lvshaoping007) [reply package.json:45]: yes , i find that isAddress different , `ethereumjs-util` just check length and hex , `web3-utils` also check sum , so i use web3 isAddress

[ethereumjs-util checkAddress](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/util/src/account.ts#L128)
[web3-utils checkAddress](https://github.com/ChainSafe/web3.js/blob/1.x/packages/web3-utils/src/utils.js#L85)

File: src/background/service/APIService.js
Comment (lvshaoping007) [reply src/background/service/APIService.js:311]: oh , is there any problem here ？

File: package.json
Comment (lvshaoping007) [reply package.json:17]: Ok , thanks for the suggestion

File: src/popup/component/AccountIcon/index.js
Comment (lvshaoping007) [reply src/popup/component/AccountIcon/index.js:17]: Here is the incoming evmAddress from the outside. Let me fix it

File: src/background/service/APIService.js
Comment (pro-wh) [reply src/background/service/APIService.js:311]: no. thanks for asking

File: src/background/api/index.js
Comment (pro-wh) [reply src/background/api/index.js:209]: the entry may be absent if the account doesn't have any tokens of that denomination, and zero in that case would be appropriate

File: src/background/api/index.js
Comment (lvshaoping007) [reply src/background/api/index.js:219]: better , i will fix this

**Comment by lvshaoping007** [comment]
> couple of notes from testing a local build:
> 
> 1. clicking the button to copy the ethereum compatible address copies the oasis address instead
> 2. the jazzicon on an imported account doesn't match metamask's rendition of the same account. seems not critical, but I can't help but look into it

> Ok I've tried doing a withdraw (and deposit again) on testnet. Some observations:
> 
> 3. I was switching around accounts to get all the addresses I need copied, and at one point the paratime tab showed "unknown" for the paratime names
> 
> And some non-critical ones:
> 
> 4. We should put a space between the paratime name and the paratime ID
> 5. It would be convenient if you could copy the evm address from the wallet tab, without having to go into the account details screen. you can currently only copy the oasis address this way.
> 6. The balance of paratime accounts is available when you go to withdraw. It might be nice to have it in the paratime tab too. But the Oasis side would probably like to discuss the design of this first.

ok , i will fix the runtime-ui and homepage copy first

File: src/popup/pages/Send/index.js
Comment (lvshaoping007) [reply src/popup/pages/Send/index.js:253]: i try catch error in [getRuntimeBalance](https://github.com/oasisprotocol/oasis-wallet-ext/blob/bdd35cc9fddd4a859009322561baec74c5fe093d/src/background/api/index.js#L207) . so there may not throw error .

File: src/background/api/index.js
Comment (lvshaoping007) [reply src/background/api/index.js:209]: yes , return 0  when balance request failed

File: src/popup/pages/Send/index.js
Comment (lvshaoping007) [reply src/popup/pages/Send/index.js:233]: i should add  await  like `await this.fetchParatimeData()` , withdraw need get runtime balance first

File: src/background/service/APIService.js
Comment (lvshaoping007) [reply src/background/service/APIService.js:208]: This suggestion is good, I will fix it

File: src/popup/pages/AccountManage/index.js
Comment (lvshaoping007) [reply src/popup/pages/AccountManage/index.js:162]: In fact , it is unique in map , but i will change key to address

File: src/background/service/APIService.js
Comment (lvshaoping007) [reply src/background/service/APIService.js:751]: oh ， dave is [oasis-sdk ](https://github.com/oasisprotocol/oasis-sdk/blob/main/client-sdk/ts-web/rt/playground/src/consensus.js#L128) , i forget change this

File: src/background/api/txHelper.js
Comment (lvshaoping007) [reply src/background/api/txHelper.js:309]: but when params.feeGas is "" . will use "" . and there  0 is Illegal in send page

File: src/popup/pages/AccountManage/index.js
Comment (lvshaoping007) [reply src/popup/pages/AccountManage/index.js]: i fix in this[ commit ](https://github.com/oasisprotocol/oasis-wallet-ext/pull/168/commits/6d3b5745c9c1a145080937165b62678b9ba4403a)

File: package.json
Comment (lvshaoping007) [reply package.json:17]: i commit yarn in this [commit ](https://github.com/oasisprotocol/oasis-wallet-ext/pull/168/commits/6b61dcc23be6062b1fc6e14c3a166fc6c632cea0)

File: src/popup/pages/AccountInfo/index.js
Comment (lvshaoping007) [reply src/popup/pages/AccountInfo/index.js]: i fix this in this [commit ](https://github.com/oasisprotocol/oasis-wallet-ext/pull/168/commits/879a33ac51f317c0af846fad4c379c06883f5a43)

File: src/popup/component/AccountIcon/index.js
Comment (lvshaoping007) [reply src/popup/component/AccountIcon/index.js:17]: i fix evm pass in[ there ](https://github.com/oasisprotocol/oasis-wallet-ext/pull/168/commits/ec72bb02cc866b69448f5cfb94c133b1ee406fd4)

File: src/background/api/index.js
Comment (lvshaoping007) [reply src/background/api/index.js]: there i commit twice , because i find  `""` will pass in  `??` . so we need check is propDecimals a number .
the last commit is [this](https://github.com/oasisprotocol/oasis-wallet-ext/pull/168/commits/deaecedcd502df8d180ccbeb47b72d8d353cd194)

File: src/popup/pages/Send/index.js
Comment (lvshaoping007) [reply src/popup/pages/Send/index.js]: fix in [this commit](https://github.com/oasisprotocol/oasis-wallet-ext/pull/168/commits/ab7084734c1de5976083f7c54fd728a0b2b2b891)

File: src/background/service/APIService.js
Comment (lvshaoping007) [reply src/background/service/APIService.js:210]: [fix account sort ](https://github.com/oasisprotocol/oasis-wallet-ext/pull/168/commits/0c7a7309051fa28885b4445916915470bca42011)

File: src/popup/pages/AccountManage/index.js
Comment (lvshaoping007) [reply src/popup/pages/AccountManage/index.js]: change accountItem key in [there](https://github.com/oasisprotocol/oasis-wallet-ext/pull/168/commits/8cf0ca300b35cbd8f9e2f08af8e2140df3dd0496)

File: src/background/service/APIService.js
Comment (lvshaoping007) [reply src/background/service/APIService.js]: fix dave in this [commit](https://github.com/oasisprotocol/oasis-wallet-ext/pull/168/commits/8c7caac1389d01433fff9c0ff5f2a121845dcece)

File: src/background/service/APIService.js
Comment (pro-wh) [reply src/background/service/APIService.js:371]: gonna resolve this

File: src/background/service/APIService.js
Comment (pro-wh) [reply src/background/service/APIService.js]: thanks! resolving

File: src/background/api/index.js
Comment (pro-wh) [reply src/background/api/index.js:209]: oh correction about my comment above: I think Luka was talking about the `.catch(...)` causing `balancesResult.balances` to be undefined. That's not equivalent to having zero balance

File: src/popup/pages/Send/index.js
Comment (pro-wh) [reply src/popup/pages/Send/index.js:253]: Ah yes. @lukaw3d, the code style of this repo is very much away from using throws and promise rejections for errors. In fact the control flow seems to be designed to diverge as little as possible, and the codebase instead substitutes special values to indicate errors. Usually responses will be substituted with an error object or numbers will hold a zero value. The latter does have the potential to create confusion though, as zero seems more to say "you lost all your funds" than "your wifi lost connection." But that is indeed the state of the codebase at this time.

File: src/popup/component/AccountIcon/index.js
Comment (pro-wh) [reply src/popup/component/AccountIcon/index.js:17]: confirmed, thanks!

File: src/popup/pages/AccountInfo/index.js
Comment (pro-wh) [reply src/popup/pages/AccountInfo/index.js]: :+1: resolving

File: package.json
Comment (pro-wh) [reply package.json:17]: got it, thanks

File: src/popup/pages/AccountManage/index.js
Comment (pro-wh) [reply src/popup/pages/AccountManage/index.js]: confirmed, resolving

File: src/popup/pages/Send/index.js
Comment (lukaw3d) [reply src/popup/pages/Send/index.js]: Also `await this.fetchData()`

File: src/popup/pages/Send/index.js
Comment (lukaw3d) [reply src/popup/pages/Send/index.js:253]: :shrug: okay

There should be `this.isRequest = false` at the end, at least?

File: src/background/api/txHelper.js
Comment (lukaw3d) [reply src/background/api/txHelper.js:309]: Ah. I'll add a comment :)
```suggestion
    // Use default if feeGas is "" or 0 (0 is illegal in send page)
    let feeGas = params.feeGas||50000
```

File: src/popup/pages/Send/index.js
Comment (lvshaoping007) [reply src/popup/pages/Send/index.js:253]: yes， i should add

File: src/popup/pages/Send/index.js
Comment (lvshaoping007) [reply src/popup/pages/Send/index.js]: ok

File: src/popup/pages/Send/index.js
Comment (lvshaoping007) [reply src/popup/pages/Send/index.js]: i fix at [this commit ](https://github.com/oasisprotocol/oasis-wallet-ext/pull/168/commits/515758afed2a67235c09a00c9420a0d4d1da8d48)

File: src/popup/pages/Send/index.js
Comment (lvshaoping007) [reply src/popup/pages/Send/index.js]: yes , i add at[ this commit](https://github.com/oasisprotocol/oasis-wallet-ext/pull/168/commits/515758afed2a67235c09a00c9420a0d4d1da8d48)

File: src/popup/pages/Send/index.js
Comment (lukaw3d) [inline src/popup/pages/Send/index.js:290]: Missing await?

**Comment by pro-wh** [comment]
Got some feedback from out side's testing with a local build:

Could we have it show the paratime transaction hash and oasisscan link when it finishes a deposit/withdraw?

**Comment by sugarT365** [comment]
> Got some feedback from out side's testing with a local build:
> 
> Could we have it show the paratime transaction hash and oasisscan link when it finishes a deposit/withdraw?
> 
> Currently depositing creates an 'Allow' entry in the transaction history, which is a separate part of the overall observability of these new operations.
> 
> edit: wait, does it already? there seems to be code already in `submitEvmWithdrawToConsensusAccountTx` and `submitParatimeAccountTx`. I'll check on my copy
> 
> edit: there is a very brief notification that links to the transaction. I'll ask if this is okay
> 
> deposit-confirm

do you mean Notification? Now when the transaction is confirmed on the chain, chrome-notification will pop up, and the user can see the notification and click to jump to oasisscan-transaction-link? If this is the case, the `withdraw and deposit of emerald` have been Added

**Comment by lvshaoping007** [comment]
> Got some feedback from out side's testing with a local build:
> 
> Could we have it show the paratime transaction hash and oasisscan link when it finishes a deposit/withdraw?
> 
> Currently depositing creates an 'Allow' entry in the transaction history, which is a separate part of the overall observability of these new operations.
> 
> edit: wait, does it already? there seems to be code already in `submitEvmWithdrawToConsensusAccountTx` and `submitParatimeAccountTx`. I'll check on my copy
> 
> edit: there is a very brief notification that links to the transaction. I'll ask if this is okay
> 
> deposit-confirm

do you mean Notification? Now when the transaction is confirmed on the chain, chrome-notification will pop up, and the user can see the notification and click to jump to oasisscan-transaction-link? If this is the case, the withdraw and deposit of emerald have been Added

File: src/popup/pages/Paratime/index.js
Comment (lukaw3d) [inline src/popup/pages/Paratime/index.js:35]: I sometimes get "unknown" as runtime names
Peek 2021-12-10 04-21

File: src/popup/pages/Send/index.js
Comment (lvshaoping007) [reply src/popup/pages/Send/index.js:290]: no , there does not need await , just silent update ，if use await here . may cause block process .

File: src/popup/pages/Paratime/index.js
Comment (lvshaoping007) [reply src/popup/pages/Paratime/index.js:35]: The runtime names returned by the network are used here. So there will be this situation, because grpc will continue to request runtime-names after returning to the list. When the runtime-names request is returned, it will be matched to display the correct name instead of the default unknown.
The purpose of using the interface to return the name here is to get the latest name in time 

https://github.com/oasisprotocol/oasis-wallet-ext/blob/121c3be81798b605bf548b9ec8893202e7ff5c81/src/popup/pages/Paratime/index.js#L94

**Review by pro-wh — APPROVED** [review]
Hi, thanks for the changes fixes so far. We'd like to get this merged into master and split of additional work new PRs.

My comments are addressed and the remaining conversations are finishing up, so I'm going to enter an approval at this time.

Other reviewers, please continue to highlight any issues that should block merging in this PR, but please also shift to filing new issues for things that would be okay to work on after merging this.

We believe that getting this merged and handling the remaining work in new branches and PRs will make it easier for other contributors from the Oasis side to help with some of the changes.

Our lint indicates that the following are not in its dictionary:

- xlink
- linecap
- linejoin
- polyline

These all appear to be correctly spelled SVG attributes, so please add these to the dictionary if it's needed to get this merged.

File: src/popup/pages/Paratime/index.js
Comment (lukaw3d) [reply src/popup/pages/Paratime/index.js:35]: It looks like you need either
```diff
         getRuntimeNameList().then((data) => {
             this.props.updateRuntimeNameList(data)
+            this.callSetState({
+                currentShowList: this.getShowRuntimeList()
+            })
         })
```

or
```diff
-  getRuntimeNameList().then((data) => {
-      this.props.updateRuntimeNameList(data)
-  })
+  let names = getRuntimeNameList()
   let rpcBalance = getRpcBalance(currentAccount.address)
   let runtimeRequestList = getRpcRuntimeList()
-  Promise.all([rpcBalance, runtimeRequestList]).then(async (data) => {
+  Promise.all([rpcBalance, runtimeRequestList, names]).then(async (data) => {
+      this.props.updateRuntimeNameList(data[2])
       let rpcAccount = data[0]
```

Comment (lvshaoping007) [reply src/popup/pages/Paratime/index.js:35]: i change name and runtimeId match from net-data to local-data . It’s faster and won’t go wrong

**Comment by lvshaoping007** [comment]
> Hi, thanks for the changes and fixes so far. We'd like to get this merged into master and split of additional work new PRs.
> 
> My comments are addressed and the remaining conversations are finishing up, so I'm going to enter an approval at this time.
> 
> Other reviewers, please continue to highlight any issues that should block merging in this PR, but please also shift to filing new issues for things that would be okay to work on after merging this.
> 
> We believe that getting this merged and handling the remaining work in new branches and PRs will make it easier for other contributors from the Oasis side to help with some of the changes.
> 
> Our lint indicates that the following are not in its dictionary:
> 
> * xlink
> * linecap
> * linejoin
> * polyline
> 
> These all appear to be correctly spelled SVG attributes, so please add these to the dictionary if it's needed to get this merged.

ok , i add cspell svg terms in the [new commit](https://github.com/oasisprotocol/oasis-wallet-ext/pull/168/commits/e90efc136f15f8b70d89f7fc6b7a36d776273309)

**Comment by pro-wh** [comment]
Thanks. Going to squash and merge

**Pull request merged**
